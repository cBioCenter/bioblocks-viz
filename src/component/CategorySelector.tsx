import * as React from 'react';
import { Dropdown } from 'semantic-ui-react';

export interface ICategorySelectorProps {
  /** Categories to select from. */
  categories: string[];
  /** Callback for when a new category is selected. */
  onCategoryChange?: (event: React.SyntheticEvent<any>, data: object) => void;
}

/**
 * Class to represent a dropdown.
 *
 * @extends {React.Component<ICategorySelectorProps, any>}
 */
export default class CategorySelector extends React.Component<ICategorySelectorProps, any> {
  constructor(props: ICategorySelectorProps) {
    super(props);
  }

  public render() {
    return (
      this.props.categories && (
        <Dropdown
          fluid={true}
          onChange={this.props.onCategoryChange}
          options={[
            { key: 'all', text: 'all', value: undefined },
            ...this.props.categories.map(cat => {
              return { key: cat, text: cat, value: cat };
            }),
          ]}
          placeholder={'Select Category'}
          search={true}
        />
      )
    );
  }
}

export { CategorySelector };
