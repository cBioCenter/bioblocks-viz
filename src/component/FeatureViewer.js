"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var React = require("react");
var immutable_1 = require("immutable");
var component_1 = require("~bioblocks-viz~/component");
var data_1 = require("~bioblocks-viz~/data");
var FeatureViewer = /** @class */ (function (_super) {
    tslib_1.__extends(FeatureViewer, _super);
    function FeatureViewer(props) {
        var _this = _super.call(this, props) || this;
        _this.onFeatureHover = function (event) {
            var e_1, _a;
            var _b = _this.props, data = _b.data, getTextForHover = _b.getTextForHover;
            var hoveredFeatureIndex = -1;
            // TODO Handle vertical viewer, better selection logic.
            var xCoords = [event.selectedPoints[0], event.selectedPoints[2]];
            for (var i = 0; i < data.length; ++i) {
                try {
                    for (var xCoords_1 = (e_1 = void 0, tslib_1.__values(xCoords)), xCoords_1_1 = xCoords_1.next(); !xCoords_1_1.done; xCoords_1_1 = xCoords_1.next()) {
                        var xCoord = xCoords_1_1.value;
                        if (data[i].contains(xCoord)) {
                            hoveredFeatureIndex = i;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (xCoords_1_1 && !xCoords_1_1.done && (_a = xCoords_1.return)) _a.call(xCoords_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            _this.setState({
                hoverAnnotationText: getTextForHover && hoveredFeatureIndex >= 0
                    ? getTextForHover(data[hoveredFeatureIndex].label, hoveredFeatureIndex)
                    : '',
                hoveredFeatureIndex: hoveredFeatureIndex,
            });
        };
        _this.onFeatureClick = function (event) {
            var _a = _this.props, data = _a.data, onClickCallback = _a.onClickCallback;
            var selectedFeatureIndices = _this.deriveFeatureIndices(data, event.selectedPoints);
            if (onClickCallback) {
                onClickCallback(_this.deriveSelectedFeatures(data, selectedFeatureIndices.toArray()));
            }
            _this.setState({
                selectedFeatureIndices: selectedFeatureIndices,
                selectedRange: new data_1.Bioblocks1DSection('selection', -1, -1),
            });
        };
        _this.onFeatureSelect = function (event) {
            var _a = _this.props, data = _a.data, onSelectCallback = _a.onSelectCallback;
            var selectedFeatureIndices = _this.deriveFeatureIndices(data, event.selectedPoints);
            var plotlyEvent = event.plotlyEvent;
            var selectedRange = _this.state.selectedRange;
            if (plotlyEvent.range) {
                selectedRange = new data_1.Bioblocks1DSection('selection', Math.floor(plotlyEvent.range.x[0]), Math.ceil(plotlyEvent.range.x[1]));
                var selection = {
                    end: selectedRange.end,
                    featuresSelected: _this.deriveSelectedFeatures(data, selectedFeatureIndices.toArray()),
                    length: selectedRange.length,
                    start: selectedRange.start,
                };
                if (onSelectCallback) {
                    onSelectCallback(selection);
                }
            }
            _this.setState({
                selectedFeatureIndices: selectedFeatureIndices,
                selectedRange: selectedRange,
            });
        };
        /**
         * Derive the indices of the Features from the points the user selected.
         */
        _this.deriveFeatureIndices = function (data, userSelectedPoints) {
            var featureIndices = immutable_1.Set();
            var _loop_1 = function (i) {
                var xCoord = userSelectedPoints[i];
                data
                    .reduce(function (result, datum, index) { return (datum.contains(xCoord) ? tslib_1.__spread(result, [index]) : result); }, new Array())
                    .forEach(function (indexToAdd) {
                    featureIndices = featureIndices.add(indexToAdd);
                });
            };
            // Points come to us as [x0, y0, x1, y1, ..., xn, yn], so we skip every other point.
            for (var i = 0; i < userSelectedPoints.length; i += 2) {
                _loop_1(i);
            }
            return featureIndices;
        };
        /**
         * Shorthand to get the raw section data for a set of Features given some indices.
         */
        _this.deriveSelectedFeatures = function (data, selectedFeatureIndices) {
            return selectedFeatureIndices.map(function (index) { return new data_1.Bioblocks1DSection(data[index].label, data[index].start, data[index].end); });
        };
        _this.state = {
            hoverAnnotationText: '',
            hoveredFeatureIndex: -1,
            plotlyConfig: {
                showAxisDragHandles: false,
                showAxisRangeEntryBoxes: false,
            },
            plotlyData: [],
            plotlyLayout: {},
            selectedFeatureIndices: immutable_1.Set(),
            selectedRange: new data_1.Bioblocks1DSection('selection', -1, -1),
        };
        return _this;
    }
    FeatureViewer.getDerivedStateFromProps = function (nextProps, nextState) {
        var backgroundBar = nextProps.backgroundBar, data = nextProps.data, height = nextProps.height, maxLength = nextProps.maxLength, showGrouped = nextProps.showGrouped, title = nextProps.title, width = nextProps.width;
        var hoverAnnotationText = nextState.hoverAnnotationText, hoveredFeatureIndex = nextState.hoveredFeatureIndex, selectedRange = nextState.selectedRange;
        var maxGroups = -1;
        var groupData = data.map(function (datum, index) {
            var yIndex = showGrouped
                ? index
                : data.findIndex(function (candidateDatum) { return datum.label.localeCompare(candidateDatum.label) === 0; });
            maxGroups = Math.max(maxGroups, yIndex);
            return FeatureViewer.getPlotlyDataObject(datum, showGrouped, yIndex);
        });
        var backgroundBarData = FeatureViewer.getPlotlyBackgroundBarObject(backgroundBar
            ? backgroundBar
            : new data_1.TintedBioblocks1DSection('', 0, data.reduce(function (prev, cur) { return Math.max(prev, cur.end); }, -1), '#b9bcb6'), showGrouped, maxGroups / 2);
        var selectionData = {
            fill: 'toself',
            hoverinfo: 'none',
            line: {
                width: 0,
            },
            marker: {
                color: '#f9f69f',
                opacity: 0.5,
            },
            mode: 'lines+markers',
            showlegend: false,
            x: selectedRange.start !== -1 && selectedRange.end !== -1
                ? [selectedRange.start, selectedRange.end, selectedRange.end, selectedRange.start]
                : [],
            y: [0, 0, maxGroups + 1, maxGroups + 1],
        };
        var plotlyData = tslib_1.__spread([backgroundBarData, selectionData], groupData);
        return {
            plotlyData: plotlyData,
            plotlyLayout: {
                annotations: FeatureViewer.getAnnotationPlotlyData(hoveredFeatureIndex, hoverAnnotationText, plotlyData[hoveredFeatureIndex]),
                dragmode: 'select',
                height: height,
                hovermode: 'closest',
                margin: {
                    b: 30,
                    t: 60,
                },
                showlegend: false,
                title: title,
                width: width,
                xaxis: data.length > 0
                    ? {
                        autorange: false,
                        fixedrange: true,
                        range: [0, maxLength ? maxLength : data.reduce(function (prev, cur) { return Math.max(prev, cur.end); }, -1) + 200],
                        showgrid: false,
                        tick0: 0,
                        tickmode: 'auto',
                        ticks: 'outside',
                    }
                    : { visible: false },
                yaxis: {
                    autorange: false,
                    fixedrange: true,
                    range: [-0.25, showGrouped ? 2 : data.length],
                    visible: false,
                },
            },
        };
    };
    FeatureViewer.getBoxForBioblocksSection = function (datum) {
        return [
            datum.end - (datum.end - datum.start) / 2,
            null,
            datum.start,
            datum.start,
            datum.end,
            datum.end,
            datum.start,
        ];
    };
    FeatureViewer.prototype.render = function () {
        var _a = this.props, width = _a.width, height = _a.height;
        var _b = this.state, plotlyConfig = _b.plotlyConfig, plotlyData = _b.plotlyData, plotlyLayout = _b.plotlyLayout;
        return (React.createElement("div", { style: { height: height, width: width } },
            React.createElement(component_1.PlotlyChart, { config: plotlyConfig, data: plotlyData, layout: plotlyLayout, onClickCallback: this.onFeatureClick, onHoverCallback: this.onFeatureHover, onSelectedCallback: this.onFeatureSelect, showLoader: false })));
    };
    FeatureViewer.defaultProps = {
        data: [],
        height: 200,
        showGrouped: false,
        title: '',
        width: 600,
    };
    FeatureViewer.getAnnotationPlotlyData = function (hoveredFeatureIndex, text, hoveredDatum) {
        return hoveredFeatureIndex >= 0 && hoveredDatum.x && hoveredDatum.y
            ? [
                {
                    align: 'left',
                    arrowhead: 0,
                    arrowsize: 1,
                    arrowwidth: 1,
                    ax: 0,
                    ay: -25,
                    bgcolor: '#ffffff',
                    bordercolor: '#000000',
                    borderpad: 5,
                    showarrow: true,
                    text: text,
                    x: hoveredDatum.x[0],
                    xref: 'x',
                    y: hoveredDatum.y[hoveredDatum.y.length - 3],
                    yref: 'y',
                },
            ]
            : [];
    };
    FeatureViewer.getPlotlyBackgroundBarObject = function (datum, showGrouped, yIndex) { return (tslib_1.__assign(tslib_1.__assign({}, FeatureViewer.getPlotlyDataObject(datum, showGrouped, yIndex)), { y: showGrouped
            ? [0.25, null, 0.25, 0.75, 0.75, 0.25, 0.25]
            : [yIndex + 0.5, null, yIndex + 1, yIndex, yIndex, yIndex + 1, yIndex + 1] })); };
    FeatureViewer.getPlotlyDataObject = function (datum, showGrouped, yIndex) { return ({
        fill: 'toself',
        fillcolor: datum.color.toString(),
        hoverinfo: 'none',
        hoveron: 'fills',
        line: {
            width: 0,
        },
        mode: 'text+lines',
        name: "" + datum.label,
        text: [datum.label],
        textfont: { color: ['#FFFFFF'] },
        type: 'scatter',
        // Creates a 'box' so we can fill it and hover over it and add a point to the middle for the label.
        x: FeatureViewer.getBoxForBioblocksSection(datum),
        y: showGrouped
            ? [0.5, null, 0, 1, 1, 0, 0]
            : [yIndex + 0.5, null, yIndex + 1, yIndex, yIndex, yIndex + 1, yIndex + 1],
    }); };
    return FeatureViewer;
}(React.Component));
exports.FeatureViewer = FeatureViewer;
//# sourceMappingURL=FeatureViewer.js.map