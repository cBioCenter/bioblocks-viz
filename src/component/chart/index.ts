export * from './AuxiliaryAxis';
export * from './ContactMapChart';
export * from './PlotlyChart';
export * from './SecondaryStructureAxis';
