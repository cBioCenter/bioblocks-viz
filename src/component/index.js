"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./chart"), exports);
tslib_1.__exportStar(require("./widget"), exports);
tslib_1.__exportStar(require("./ContactMapComponent"), exports);
tslib_1.__exportStar(require("./FeatureViewer"), exports);
tslib_1.__exportStar(require("./NGLComponent"), exports);
tslib_1.__exportStar(require("./TensorTComponent"), exports);
tslib_1.__exportStar(require("./UMAPVisualization"), exports);
//# sourceMappingURL=index.js.map