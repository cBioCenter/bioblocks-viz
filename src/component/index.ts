export * from './chart';
export * from './widget';

export * from './ContactMap';
export * from './FeatureViewer';
export * from './InfoPanel';
export * from './MorpheusComponent';
export * from './NGLComponent';
export * from './PredictedContactMap';
export * from './TComponent';
export * from './TensorTComponent';
export * from './TFrameComponent';
export * from './VizSelectorPanel';
