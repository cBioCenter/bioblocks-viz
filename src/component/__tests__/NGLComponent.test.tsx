import { mount, ReactWrapper, shallow } from 'enzyme';
import * as NGL from 'ngl';
import * as React from 'react';

import { NGLComponent, NGLComponentClass } from '~chell-viz~/component';
import { initialResidueContext, initialSecondaryStructureContext } from '~chell-viz~/context';

describe('NGLComponent', () => {
  let sampleData: NGL.Structure;

  beforeEach(() => {
    sampleData = new NGL.Structure();
    if (!sampleData.residueStore) {
      sampleData.residueStore = new NGL.ResidueStore();
    }
    sampleData.residueStore.resno = new Uint32Array([0, 1, 2, 3, 4]);
  });

  it('Should match existing snapshot when unconnected to a context.', () => {
    expect(shallow(<NGLComponentClass />)).toMatchSnapshot();
  });

  it('Should handle prop updates.', () => {
    const wrapper = mount(<NGLComponent />);
    const initialProps = wrapper.props();

    wrapper.setProps({
      hoveredResidues: [0],
    });

    expect(wrapper.props()).not.toEqual(initialProps);
  });

  it('Should handle candidate residue updates.', () => {
    const wrapper = mount(<NGLComponent data={sampleData} />);

    wrapper.setProps({
      candidateResidues: [1, 2, 3],
    });
  });

  it('Should handle hovered residue updates.', () => {
    const wrapper = mount(<NGLComponent data={sampleData} />);

    wrapper.setProps({
      hoveredResidues: [1, 2, 3],
    });
  });

  it('Should show the ball+stick representation for hovered residues.', () => {
    const expectedRep = 'ball+stick';
    const wrapper = mount(<NGLComponentClass data={sampleData} />);
    const instance = wrapper.instance() as NGLComponentClass;

    wrapper.setProps({
      residueContext: {
        ...initialResidueContext,
        hoveredResidues: [1],
      },
    });

    wrapper.update();

    expect(instance.state.activeRepresentations[0].name()).toEqual(expectedRep);
  });

  it('Should show the distance and ball+stick representation for locked residues.', () => {
    const expectedRep = ['ball+stick', 'distance', 'ball+stick', 'distance'];
    const wrapper = mount(<NGLComponentClass data={sampleData} />);
    const instance = wrapper.instance() as NGLComponentClass;

    wrapper.setProps({
      residueContext: {
        ...initialResidueContext,
        lockedResiduePairs: new Map(
          Object.entries({
            '1,2': [1, 2],
            '3,4': [3, 4],
          }),
        ),
      },
    });

    const { activeRepresentations } = instance.state;
    for (let i = 0; i < activeRepresentations.length; i++) {
      expect(activeRepresentations[i].name()).toEqual(expectedRep[i]);
    }
  });

  it('Should show the distance and ball+stick representation for multiple hovered residues.', () => {
    const expectedRep = ['ball+stick', 'distance'];
    const wrapper = mount(<NGLComponentClass data={sampleData} />);
    const instance = wrapper.instance() as NGLComponentClass;

    wrapper.setProps({
      residueContext: {
        ...initialResidueContext,
        hoveredResidues: [7, 8],
      },
    });

    wrapper.update();

    const { activeRepresentations } = instance.state;
    expect(activeRepresentations.length).toEqual(expectedRep.length);
    for (let i = 0; i < activeRepresentations.length; i++) {
      expect(activeRepresentations[i].name()).toEqual(expectedRep[i]);
    }
  });

  it('Should show the cartoon representation for selected secondary structures.', () => {
    const expectedRep = ['cartoon'];
    const wrapper = mount(<NGLComponentClass data={sampleData} />);
    const instance = wrapper.instance() as NGLComponentClass;

    wrapper.setProps({
      secondaryStructureContext: {
        ...initialSecondaryStructureContext,
        selectedSecondaryStructures: [{ start: 1, end: 2 }],
      },
    });

    const { activeRepresentations } = instance.state;
    expect(activeRepresentations.length).toEqual(expectedRep.length);
    for (let i = 0; i < activeRepresentations.length; i++) {
      expect(activeRepresentations[i].name()).toEqual(expectedRep[i]);
    }
  });

  it('Should follow candidate selection flow.', () => {
    const wrapper = mount(<NGLComponentClass data={sampleData} />);

    wrapper.setProps({
      lockedResiduePairs: new Map(
        Object.entries({
          '1,2': [1, 2],
        }),
      ),
    });
    wrapper.update();

    wrapper.setProps({
      lockedResiduePairs: new Map(
        Object.entries({
          '1,2': [1, 2],
          '3,4': [3, 4],
        }),
      ),
    });
    wrapper.update();
  });

  it('Should follow candidate residue selection flow.', () => {
    const wrapper = mount(<NGLComponentClass data={sampleData} />);

    wrapper.setProps({
      candidateResidues: [1],
    });
    wrapper.update();

    wrapper.setProps({
      candidateResidues: [2],
    });
    wrapper.update();
  });

  it('Should call appropriate residue clearing callback.', () => {
    const removeSpy = jest.fn();
    const wrapper = mount(
      <NGLComponentClass residueContext={{ ...initialResidueContext, removeAllLockedResiduePairs: removeSpy }} />,
    );
    wrapper
      .find('#remove-all-locked-distance-pairs-0')
      .at(0)
      .simulate('click');
    expect(removeSpy).toHaveBeenCalledTimes(1);
  });

  it('Should unmount correctly.', () => {
    const wrapper = mount(<NGLComponentClass />);
    expect(wrapper.get(0)).not.toBeNull();
    wrapper.unmount();
    expect(wrapper.get(0)).toBeUndefined();
  });

  describe('Events', () => {
    const simulateHoverEvent = (wrapper: ReactWrapper<any, any>, opts: object) => {
      const instance = wrapper.instance() as NGLComponentClass;
      instance.state.stage!.mouseControls.run(NGL.MouseActions.HOVER_PICK, instance.state.stage, opts);
    };

    const simulateClickEvent = (wrapper: ReactWrapper<any, any>, opts?: object) => {
      const instance = wrapper.instance() as NGLComponentClass;
      instance.state.stage!.signals.clicked.dispatch(opts);
    };

    it('Should handle hover events when there is no hovered or candidate residue.', async () => {
      const wrapper = mount(<NGLComponentClass data={sampleData} />);

      const removeHoveredResiduesSpy = jest.fn();
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          candidateResidues: [],
          removeHoveredResidues: removeHoveredResiduesSpy,
        },
      });
      simulateHoverEvent(wrapper, { atom: { resno: 3 } });
      simulateHoverEvent(wrapper, { closestBondAtom: { resno: 3 } });
      simulateHoverEvent(wrapper, {});
      expect(removeHoveredResiduesSpy).toHaveBeenCalledTimes(0);
    });

    it('Should handle hover events when there is a hovered residue but no candidate.', async () => {
      const wrapper = mount(<NGLComponentClass data={sampleData} />);

      const removeHoveredResiduesSpy = jest.fn();
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          candidateResidues: [],
          hoveredResidues: [3],
          removeHoveredResidues: removeHoveredResiduesSpy,
        },
      });
      simulateHoverEvent(wrapper, { atom: { resno: 3 } });
      simulateHoverEvent(wrapper, {});
      expect(removeHoveredResiduesSpy).toHaveBeenCalledTimes(1);
    });

    it('Should clear candidate and hovered residues when the mouse leaves the canvas.', async () => {
      const wrapper = mount(<NGLComponentClass data={sampleData} />);
      const removeNonLockedResiduesSpy = jest.fn();

      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          candidateResidues: [3],
          hoveredResidues: [4],
          removeNonLockedResidues: removeNonLockedResiduesSpy,
        },
      });

      wrapper.find('.NGLCanvas').simulate('mouseleave');
      expect(removeNonLockedResiduesSpy).toHaveBeenCalledTimes(1);
    });

    // @ts-ignore
    // TODO Add each to official jest types - Jest is as v23 but types are for v22 so far.
    it.each([{ atom: { resno: 4 } }, { closestBondAtom: { resno: 4 } }])(
      'Should handle click events by creating a locked residue pair if there is a candidate.',
      async (pickingResult: NGL.PickingProxy) => {
        const addLockedSpy = jest.fn();
        const wrapper = mount(<NGLComponentClass data={sampleData} />);
        wrapper.setProps({
          residueContext: {
            ...initialResidueContext,
            addLockedResiduePair: addLockedSpy,
            candidateResidues: [2],
          },
        });
        simulateClickEvent(wrapper, pickingResult);
        expect(addLockedSpy).toHaveBeenCalledWith([2, 4]);
      },
    );

    // @ts-ignore
    // TODO Add each to official jest types - Jest is as v23 but types are for v22 so far.
    it.each([{ atom: { resno: 4 } }, { closestBondAtom: { resno: 4 } }])(
      'Should handle click events by creating a candidate residue when none is present.',
      async (pickingResult: NGL.PickingProxy) => {
        const addCandidateSpy = jest.fn();
        const wrapper = mount(<NGLComponentClass data={sampleData} />);
        wrapper.setProps({
          residueContext: {
            ...initialResidueContext,
            addCandidateResidues: addCandidateSpy,
          },
        });
        simulateClickEvent(wrapper, pickingResult);
        expect(addCandidateSpy).toHaveBeenCalledWith([4]);
      },
    );

    it('Should handle clicking on a distance representation.', async () => {
      const wrapper = mount(<NGLComponentClass data={sampleData} />);
      const removedLockedSpy = jest.fn();
      (wrapper.state('structureComponent') as NGL.StructureComponent).addRepresentation('distance');
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          lockedResiduePairs: new Map(
            Object.entries({
              '4,7': [4, 7],
            }),
          ),
          removeLockedResiduePair: removedLockedSpy,
        },
      });
      simulateClickEvent(wrapper, {
        distance: { atom1: { resno: 4 }, atom2: { resno: 7 } },
        picker: { type: 'distance' },
      });
      expect(removedLockedSpy).toHaveBeenLastCalledWith([4, 7]);
    });

    it('Should handle clicking off-component.', async () => {
      const expected = new Array<string>();
      const wrapper = mount(<NGLComponentClass data={sampleData} />);
      const removeNonLockedResiduesSpy = jest.fn();
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          hoveredResidues: [1],
          removeNonLockedResidues: removeNonLockedResiduesSpy,
        },
      });
      expect(wrapper.state('activeRepresentations')).not.toEqual(expected);
      simulateClickEvent(wrapper, undefined);
      expect(removeNonLockedResiduesSpy).toHaveBeenCalledTimes(1);
    });

    it('Should add the hovered residue if the user clicks within a certain distance (snapping).', async () => {
      const expected = new Array<string>();
      const wrapper = mount(<NGLComponentClass data={sampleData} />);
      const addLockedResiduePairSpy = jest.fn();
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          addLockedResiduePair: addLockedResiduePairSpy,
          candidateResidues: [0],
          hoveredResidues: [1],
        },
      });
      expect(wrapper.state('activeRepresentations')).not.toEqual(expected);
      simulateClickEvent(wrapper, undefined);
      expect(addLockedResiduePairSpy).toHaveBeenCalledTimes(1);
    });

    it('Should remove the hovered residue if the user clicks within beyond a certain distance (snapping).', async () => {
      const expected = new Array<string>();
      const wrapper = mount(<NGLComponentClass data={sampleData} />);
      const removeNonLockedResiduesSpy = jest.fn();
      wrapper.setProps({
        residueContext: {
          ...initialResidueContext,
          candidateResidues: [0],
          hoveredResidues: [51],
          removeNonLockedResidues: removeNonLockedResiduesSpy,
        },
      });
      expect(wrapper.state('activeRepresentations')).not.toEqual(expected);
      simulateClickEvent(wrapper, undefined);
      expect(removeNonLockedResiduesSpy).toHaveBeenCalledTimes(1);
    });

    it('Should handle pressing ESC.', async () => {
      const removeNonLockedResiduesSpy = jest.fn();
      const wrapper = mount(
        <NGLComponentClass
          data={sampleData}
          residueContext={{
            ...initialResidueContext,
            candidateResidues: [1],
            hoveredResidues: [2],
            removeNonLockedResidues: removeNonLockedResiduesSpy,
          }}
        />,
      );

      wrapper.find('.NGLCanvas').simulate('keyDown', { keyCode: 27 });
      expect(removeNonLockedResiduesSpy).toHaveBeenCalledTimes(1);
    });

    it('Should handle onResize events.', () => {
      const onResizeSpy = jest.fn();
      expect(() => mount(<NGLComponent onResize={onResizeSpy} />)).not.toThrow();

      global.dispatchEvent(new Event('resize'));
      expect(onResizeSpy).toHaveBeenCalledTimes(1);
    });

    it('Should remove the onResize handler when unmounted.', async () => {
      const onResizeSpy = jest.fn();
      const wrapper = mount(<NGLComponent onResize={onResizeSpy} />);
      await wrapper.unmount();
      global.dispatchEvent(new Event('resize'));
      expect(onResizeSpy).toHaveBeenCalledTimes(0);
    });
  });
});
