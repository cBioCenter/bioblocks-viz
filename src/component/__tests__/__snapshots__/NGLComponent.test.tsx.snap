// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NGLComponent Should match existing snapshot when unconnected to a context. 1`] = `
<ComponentCard
  componentName="NGL Viewer"
  frameHeight={0}
  frameWidth={0}
  headerHeight={32}
  height="525px"
  iconSrc="assets/icons/spring-icon.png"
  isFramedComponent={false}
  isFullPage={false}
  menuItems={
    Array [
      Object {
        "component": Object {
          "configs": Array [
            Object {
              "name": "Clear Selections",
              "onClick": [Function],
              "type": "BUTTON",
            },
            Object {
              "current": "CLOSEST",
              "name": "Proximity Metric",
              "onChange": [Function],
              "options": Array [
                "Closest",
                "C-Î±",
              ],
              "type": "RADIO",
            },
            Object {
              "current": "default",
              "name": "Structure Representation Type",
              "onChange": [Function],
              "options": Array [
                "Default",
                "Spacefill",
                "Backbone",
                "Cartoon",
                "Surface",
                "Tube",
              ],
              "type": "RADIO",
            },
          ],
          "name": "SIDEBAR",
          "props": Object {
            "trigger": <Icon
              as="i"
              name="setting"
            />,
          },
        },
        "description": "Settings",
      },
    ]
  }
  padding={0}
  showSettings={false}
  width="525px"
>
  <div
    className="NGLComponent"
    style={
      Object {
        "height": "90%",
        "width": "100%",
      }
    }
  >
    <Dimmer
      active={false}
    >
      <Loader />
    </Dimmer>
    <div
      className="NGLCanvas"
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="img"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
  <div
    style={
      Object {
        "bottom": "20px",
        "position": "absolute",
        "right": "20px",
        "userSelect": "none",
      }
    }
  >
    <a
      aria-pressed={false}
      onClick={[Function]}
      role="button"
    >
      Superposition: off
    </a>
  </div>
</ComponentCard>
`;
